File: e204/e204.py
#!/usr/bin/env python

"""
ProjectEuler.net Problem 204
Solved by Christian Stigen Larsen

Find x so that x^2 = 1_2_3_4_5_6_7_8_9_0 where _ is an arbitrary digit.

math.sqrt(1020304050607080900) == 1010101010.1010101
so the answer should be right after that.
"""

from math import *
import re, sys

MIN = 1020304050607080900
MAX = 1929394959697989990

x = int(sqrt(1020304050607080900))

pattern = re.compile("1.2.3.4.5.6.7.8.9.0");
def accepted(n):
  return pattern.match(str(n))

n = 0
while not accepted(x):
  if (n % 1000000) == 0:
    print x, x*x
    sys.stdout.flush()
  n += 1
  x += 1

print x, "squared is", x*x

File: e204/e204.cpp
/*
 * ProjectEuler.net Problem 204
 * Solved by Christian Stigen Larsen
 */

#include <iostream>
#include <inttypes.h>

using namespace std;

typedef uint64_t Z;

static inline Z last(const Z& n)
{
  return n % 10;
}

static inline void div10(Z& n, Z tens)
{
  while ( tens-- )
    n /= 10;
}


// is n like 1_2_3_4_5_6_7_8_9_0
static bool accepted(const Z& b)
{
  Z n = b*b;

  // last number must be zero
  if ( last(n) != 0 ) return false;

  div10(n, 2); if ( last(n) != 9 ) return false;
  div10(n, 2); if ( last(n) != 8 ) return false;
  div10(n, 2); if ( last(n) != 7 ) return false;
  div10(n, 2); if ( last(n) != 6 ) return false;
  div10(n, 2); if ( last(n) != 5 ) return false;
  div10(n, 2); if ( last(n) != 4 ) return false;
  div10(n, 2); if ( last(n) != 3 ) return false;
  div10(n, 2); if ( last(n) != 2 ) return false;
  div10(n, 2); if ( last(n) != 1 ) return false;

  return true;
}

int main()
{
  // math.sqrt(1020304050607080900) == 1010101010.1010101
  Z n = 1010101010;

  while ( !accepted(n) )
    ++n;

  cout << n << " squared is: " << endl;
  cout << (n*n) << endl;
  cout << "1_2_3_4_5_6_7_8_9_0" << endl;
}
